File: py/projects/src/figma_projects/sdk.py
"""High-level SDK for Figma Projects API."""

import asyncio
import logging
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any, AsyncIterator

from .client import FigmaProjectsClient
from .models import (
    Project,
    ProjectFile,
    TeamProjectsResponse,
    ProjectFilesResponse,
    ProjectTree,
    ProjectStatistics,
    FileSearchResult,
    BatchProjectResult,
    ExportFormat,
)
from .utils import (
    filter_files_by_name,
    sort_files_by_last_modified,
    is_recent_file,
    calculate_file_statistics,
    export_projects_to_json,
    export_projects_to_csv,
    validate_team_id,
    validate_project_id,
)
from .errors import ValidationError, NotFoundError


logger = logging.getLogger(__name__)


class FigmaProjectsSDK:
    """High-level SDK for Figma Projects API."""
    
    def __init__(
        self,
        api_token: str,
        base_url: str = "https://api.figma.com",
        requests_per_minute: int = 60,
        timeout: float = 30.0,
        max_retries: int = 3,
    ):
        """Initialize the SDK.
        
        Args:
            api_token: Figma API token
            base_url: Base URL for the API
            requests_per_minute: Rate limit for requests
            timeout: Request timeout in seconds
            max_retries: Maximum number of retries for failed requests
        """
        self.client = FigmaProjectsClient(
            api_token=api_token,
            base_url=base_url,
            requests_per_minute=requests_per_minute,
            timeout=timeout,
            max_retries=max_retries,
        )
    
    async def __aenter__(self):
        """Async context manager entry."""
        await self.client.__aenter__()
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit."""
        await self.client.__aexit__(exc_type, exc_val, exc_tb)
    
    async def close(self) -> None:
        """Close the SDK and underlying client."""
        await self.client.close()
    
    # Core API Methods
    
    async def get_team_projects(self, team_id: str) -> TeamProjectsResponse:
        """Get all projects in a team.
        
        Args:
            team_id: ID of the team
            
        Returns:
            Team projects response
            
        Raises:
            ValidationError: If team_id is invalid
            NotFoundError: If team is not found
        """
        if not validate_team_id(team_id):
            raise ValidationError("team_id", team_id, "Invalid team ID format")
        
        path = f"/v1/teams/{team_id}/projects"
        response_data = await self.client.get(path)
        
        return TeamProjectsResponse(**response_data)
    
    async def get_project_files(
        self,
        project_id: str,
        include_branch_data: bool = False,
    ) -> ProjectFilesResponse:
        """Get all files in a project.
        
        Args:
            project_id: ID of the project
            include_branch_data: Whether to include branch metadata
            
        Returns:
            Project files response
            
        Raises:
            ValidationError: If project_id is invalid
            NotFoundError: If project is not found
        """
        if not validate_project_id(project_id):
            raise ValidationError("project_id", project_id, "Invalid project ID format")
        
        path = f"/v1/projects/{project_id}/files"
        params = {}
        if include_branch_data:
            params["branch_data"] = "true"
        
        response_data = await self.client.get(path, params=params)
        
        return ProjectFilesResponse(**response_data)
    
    # Enhanced SDK Methods
    
    async def list_all_team_projects(self, team_id: str) -> List[Project]:
        """Get all projects in a team with automatic pagination.
        
        Args:
            team_id: ID of the team
            
        Returns:
            List of all projects
        """
        response = await self.get_team_projects(team_id)
        return response.projects
    
    async def list_all_project_files(
        self,
        project_id: str,
        include_branch_data: bool = False,
    ) -> List[ProjectFile]:
        """Get all files in a project with automatic pagination.
        
        Args:
            project_id: ID of the project
            include_branch_data: Whether to include branch metadata
            
        Returns:
            List of all files
        """
        response = await self.get_project_files(project_id, include_branch_data)
        return response.files
    
    async def get_project_tree(self, team_id: str) -> ProjectTree:
        """Get hierarchical project/file structure for a team.
        
        Args:
            team_id: ID of the team
            
        Returns:
            Project tree structure
        """
        team_response = await self.get_team_projects(team_id)
        projects_with_files = []
        
        # Fetch files for each project concurrently
        async def get_project_with_files(project: Project) -> Dict[str, Any]:
            try:
                files_response = await self.get_project_files(project.id)
                return {
                    "id": project.id,
                    "name": project.name,
                    "files": [
                        {
                            "key": file.key,
                            "name": file.name,
                            "thumbnail_url": file.thumbnail_url,
                            "last_modified": file.last_modified.isoformat(),
                        }
                        for file in files_response.files
                    ],
                    "file_count": len(files_response.files),
                }
            except Exception as e:
                logger.warning(f"Failed to get files for project {project.id}: {e}")
                return {
                    "id": project.id,
                    "name": project.name,
                    "files": [],
                    "file_count": 0,
                    "error": str(e),
                }
        
        # Use semaphore to limit concurrent requests
        semaphore = asyncio.Semaphore(5)
        
        async def bounded_get_project_with_files(project: Project) -> Dict[str, Any]:
            async with semaphore:
                return await get_project_with_files(project)
        
        tasks = [
            bounded_get_project_with_files(project)
            for project in team_response.projects
        ]
        
        projects_with_files = await asyncio.gather(*tasks, return_exceptions=True)
        
        # Filter out exceptions
        valid_projects = [
            p for p in projects_with_files
            if not isinstance(p, Exception)
        ]
        
        return ProjectTree(
            team_name=team_response.name,
            projects=valid_projects,
        )
    
    async def find_file_by_name(
        self,
        project_id: str,
        file_name: str,
        exact_match: bool = False,
    ) -> Optional[ProjectFile]:
        """Find a specific file in a project by name.
        
        Args:
            project_id: ID of the project
            file_name: Name of the file to find
            exact_match: Whether to match exactly or partially
            
        Returns:
            Found file or None
        """
        files = await self.list_all_project_files(project_id)
        
        if exact_match:
            for file in files:
                if file.name == file_name:
                    return file
        else:
            for file in files:
                if file_name.lower() in file.name.lower():
                    return file
        
        return None
    
    async def get_recent_files(
        self,
        project_id: str,
        limit: int = 10,
        days: int = 30,
    ) -> List[ProjectFile]:
        """Get recently modified files in a project.
        
        Args:
            project_id: ID of the project
            limit: Maximum number of files to return
            days: Number of days to consider recent
            
        Returns:
            List of recent files
        """
        files = await self.list_all_project_files(project_id)
        recent_files = [f for f in files if is_recent_file(f, days)]
        sorted_files = sort_files_by_last_modified(recent_files, ascending=False)
        
        return sorted_files[:limit]
    
    async def search_projects(self, team_id: str, query: str) -> List[Project]:
        """Search projects by name.
        
        Args:
            team_id: ID of the team
            query: Search query
            
        Returns:
            List of matching projects
        """
        projects = await self.list_all_team_projects(team_id)
        query_lower = query.lower()
        
        return [
            project for project in projects
            if query_lower in project.name.lower()
        ]
    
    async def search_files_in_project(
        self,
        project_id: str,
        query: str,
        case_sensitive: bool = False,
    ) -> List[FileSearchResult]:
        """Search files within a project.
        
        Args:
            project_id: ID of the project
            query: Search query
            case_sensitive: Whether search should be case sensitive
            
        Returns:
            List of search results with relevance scores
        """
        project_response = await self.get_project_files(project_id)
        files = filter_files_by_name(project_response.files, query, case_sensitive)
        
        results = []
        for file in files:
            # Simple relevance scoring
            score = 1.0
            if query.lower() == file.name.lower():
                score = 1.0  # Exact match
            elif file.name.lower().startswith(query.lower()):
                score = 0.8  # Starts with query
            else:
                score = 0.5  # Contains query
            
            results.append(
                FileSearchResult(
                    file=file,
                    project_id=project_id,
                    project_name=project_response.name,
                    match_score=score,
                )
            )
        
        # Sort by relevance score
        return sorted(results, key=lambda r: r.match_score, reverse=True)
    
    async def get_project_statistics(self, project_id: str) -> ProjectStatistics:
        """Get statistics for a project.
        
        Args:
            project_id: ID of the project
            
        Returns:
            Project statistics
        """
        project_response = await self.get_project_files(project_id)
        files = project_response.files
        
        recent_files = [f for f in files if is_recent_file(f)]
        last_activity = max((f.last_modified for f in files), default=None)
        
        return ProjectStatistics(
            project_id=project_id,
            project_name=project_response.name,
            total_files=len(files),
            recent_files=len(recent_files),
            last_activity=last_activity,
        )
    
    async def batch_get_projects(self, project_ids: List[str]) -> List[BatchProjectResult]:
        """Get multiple projects at once.
        
        Args:
            project_ids: List of project IDs
            
        Returns:
            List of batch results
        """
        results = []
        
        async def get_single_project(project_id: str) -> BatchProjectResult:
            try:
                response = await self.get_project_files(project_id)
                project = Project(id=project_id, name=response.name)
                return BatchProjectResult(
                    project_id=project_id,
                    success=True,
                    project=project,
                )
            except Exception as e:
                return BatchProjectResult(
                    project_id=project_id,
                    success=False,
                    error=str(e),
                )
        
        # Use semaphore to limit concurrent requests
        semaphore = asyncio.Semaphore(3)
        
        async def bounded_get_project(project_id: str) -> BatchProjectResult:
            async with semaphore:
                return await get_single_project(project_id)
        
        tasks = [bounded_get_project(pid) for pid in project_ids]
        results = await asyncio.gather(*tasks)
        
        return results
    
    async def export_project_structure(
        self,
        team_id: str,
        format: ExportFormat = ExportFormat.JSON,
        include_files: bool = True,
    ) -> str:
        """Export project structure as JSON or CSV.
        
        Args:
            team_id: ID of the team
            format: Export format (json or csv)
            include_files: Whether to include file data
            
        Returns:
            Exported data as string
        """
        project_tree = await self.get_project_tree(team_id)
        
        if format == ExportFormat.JSON:
            return export_projects_to_json(project_tree.projects, include_files)
        elif format == ExportFormat.CSV:
            return export_projects_to_csv(project_tree.projects)
        else:
            raise ValidationError("format", format, "Unsupported export format")
    
    # Utility Methods
    
    def get_rate_limit_info(self):
        """Get current rate limit information."""
        return self.client.get_rate_limit_info()
    
    def get_client_stats(self) -> Dict[str, Any]:
        """Get client statistics."""
        return self.client.get_stats()
    
    def reset_client_stats(self) -> None:
        """Reset client statistics."""
        self.client.reset_stats()

File: ts/projects/src/interfaces/sdk.ts
/**
 * SDK facade for figma-projects
 * Provides ergonomic API over core service layer
 */

import FigmaProjectsService from "../core/service.js";

/**
 * High-level SDK for Figma Projects API
 * Provides convenient methods for project operations
 *
 * @example
 * import { FigmaApiClient } from '@figma-api/fetch';
 * import { FigmaProjectsSDK } from 'figma-projects';
 *
 * const fetcher = new FigmaApiClient({ apiToken: process.env.FIGMA_TOKEN });
 * const sdk = new FigmaProjectsSDK({ fetcher });
 */
export class FigmaProjectsSDK {
  service: FigmaProjectsService;
  logger: any;

  /**
   * @param {object} config - SDK configuration
   * @param {object} config.fetcher - FigmaApiClient instance (required)
   * @param {object} [config.logger=console] - Logger instance
   * @param {object} [config.serviceConfig] - Service layer configuration
   */
  constructor({
    fetcher,
    logger = console,
    serviceConfig = {},
  }: {
    fetcher?: any;
    logger?: any;
    serviceConfig?: Record<string, any>;
  } = {}) {
    // Initialize service
    this.service = new FigmaProjectsService({
      fetcher,
      logger,
      config: serviceConfig,
    });

    this.logger = logger;
    this.logger.debug("FigmaProjectsSDK initialized");
  }

  /**
   * Get all projects in a team
   * @param {string} teamId - Team ID
   * @param {object} [options={}] - Options
   * @param {boolean} [options.includeStats=false] - Include project statistics
   * @returns {Promise<object>} Projects data
   */
  async getTeamProjects(
    teamId: string,
    options: Record<string, any> = {},
  ): Promise<Record<string, any>> {
    return this.service.getTeamProjects(teamId, options);
  }

  /**
   * Get all files in a project
   * @param {string} projectId - Project ID
   * @param {object} [options={}] - Options
   * @param {boolean} [options.branchData=false] - Include branch metadata
   * @param {boolean} [options.sortByModified=true] - Sort by last modified date
   * @returns {Promise<object>} Project files data
   */
  async getProjectFiles(
    projectId: string,
    options: Record<string, any> = {},
  ): Promise<Record<string, any>> {
    return this.service.getProjectFiles(projectId, options);
  }

  /**
   * Get complete project tree (projects with their files)
   * @param {string} teamId - Team ID
   * @param {object} [options={}] - Options
   * @param {number} [options.maxConcurrency=5] - Max concurrent file requests
   * @param {boolean} [options.includeEmptyProjects=true] - Include projects with no files
   * @returns {Promise<object>} Complete project tree
   */
  async getProjectTree(
    teamId: string,
    options: Record<string, any> = {},
  ): Promise<Record<string, any>> {
    return this.service.getProjectsWithFiles(teamId, options);
  }

  /**
   * Search for projects by name
   * @param {string} teamId - Team ID
   * @param {string} query - Search query
   * @param {object} [options={}] - Search options
   * @param {boolean} [options.caseSensitive=false] - Case sensitive search
   * @param {boolean} [options.exactMatch=false] - Exact match only
   * @returns {Promise<object>} Search results
   */
  async searchProjects(
    teamId: string,
    query: string,
    options: Record<string, any> = {},
  ): Promise<Record<string, any>> {
    return this.service.searchProjects(teamId, query, options);
  }

  /**
   * Get project statistics and metrics
   * @param {string} projectId - Project ID
   * @returns {Promise<object>} Project statistics
   */
  async getProjectStats(projectId: string): Promise<Record<string, any>> {
    return this.service.getProjectStatistics(projectId);
  }

  /**
   * Find a specific file by name within a project
   * @param {string} projectId - Project ID
   * @param {string} fileName - File name to search for
   * @param {object} [options={}] - Search options
   * @param {boolean} [options.caseSensitive=false] - Case sensitive search
   * @param {boolean} [options.exactMatch=true] - Exact match only
   * @returns {Promise<object>} File data
   */
  async findFile(
    projectId: string,
    fileName: string,
    options: Record<string, any> = {},
  ): Promise<Record<string, any> | null> {
    return this.service.findFileByName(projectId, fileName, options);
  }

  /**
   * Get recently modified files across all projects in a team
   * @param {string} teamId - Team ID
   * @param {number} [limit=10] - Maximum number of files to return
   * @param {number} [daysBack=7] - Number of days to look back
   * @returns {Promise<object>} Recent files data
   */
  async getRecentFiles(
    teamId: string,
    limit: number = 10,
    daysBack: number = 7,
  ): Promise<Record<string, any>> {
    return this.service.getRecentFiles(teamId, limit, daysBack);
  }

  /**
   * Get multiple projects at once
   * @param {string[]} projectIds - Array of project IDs
   * @param {object} [options={}] - Options
   * @param {number} [options.maxConcurrency=5] - Max concurrent requests
   * @returns {Promise<object>} Batch results
   */
  async getMultipleProjects(
    projectIds: string[],
    options: Record<string, any> = {},
  ): Promise<Record<string, any>> {
    return this.service.batchGetProjects(projectIds, options);
  }

  /**
   * Export project structure to different formats
   * @param {string} teamId - Team ID
   * @param {string} [format='json'] - Export format (json, csv)
   * @returns {Promise<string>} Exported data
   */
  async exportProjects(
    teamId: string,
    format: string = "json",
  ): Promise<string> {
    return this.service.exportProjectStructure(teamId, format);
  }

  /**
   * Health check - verify API connectivity and authentication
   * @returns {Promise<object>} Health status
   */
  async healthCheck(): Promise<Record<string, any>> {
    try {
      this.logger.debug("Performing health check");

      // Simple health check - SDK is operational if service is initialized
      const status = {
        status: "healthy",
        timestamp: new Date().toISOString(),
        sdkInitialized: true,
        serviceInitialized: !!this.service,
        version: "1.0.0",
      };

      this.logger.debug("Health check completed", status);
      return status;
    } catch (error: any) {
      this.logger.error("Health check failed", { error: error.message });

      return {
        status: "unhealthy",
        timestamp: new Date().toISOString(),
        error: error.message,
        code: error.code || "UNKNOWN_ERROR",
        version: "1.0.0",
      };
    }
  }

  /**
   * Get SDK metrics and performance data
   * @returns {object} Comprehensive SDK metrics
   */
  getMetrics(): Record<string, any> {
    return {
      sdkVersion: "1.0.0",
      serviceInitialized: !!this.service,
      timestamp: new Date().toISOString(),
    };
  }

  /**
   * Reset SDK state (placeholder for future cache/metrics implementation)
   */
  reset(): void {
    this.logger.debug("SDK state reset");
  }

  // Convenience methods for common workflows

  /**
   * Get a complete overview of a team's projects and activity
   * @param {string} teamId - Team ID
   * @param {object} [options={}] - Options
   * @param {boolean} [options.includeStats=true] - Include project statistics
   * @param {boolean} [options.includeRecentFiles=true] - Include recent files
   * @param {number} [options.recentFilesLimit=20] - Limit for recent files
   * @returns {Promise<object>} Team overview
   */
  async getTeamOverview(
    teamId: string,
    options: Record<string, any> = {},
  ): Promise<Record<string, any>> {
    const includeStats = options.includeStats !== false;
    const includeRecentFiles = options.includeRecentFiles !== false;
    const recentFilesLimit = options.recentFilesLimit || 20;

    try {
      this.logger.debug("Generating team overview", { teamId, options });

      // Get projects with statistics
      const projects = await this.getTeamProjects(teamId, { includeStats });

      // Get recent files if requested
      let recentFiles: any = null;
      if (includeRecentFiles) {
        try {
          recentFiles = await this.getRecentFiles(teamId, recentFilesLimit);
        } catch (error: any) {
          this.logger.warn("Failed to fetch recent files for overview", {
            teamId,
            error: error.message,
          });
        }
      }

      // Calculate summary statistics
      const totalProjects = projects.totalCount;
      const projectsWithStats = projects.projects.filter(
        (p: any) => p.statistics,
      );
      const totalFiles = projectsWithStats.reduce(
        (sum: number, p: any) => sum + (p.statistics?.fileCount || 0),
        0,
      );
      const activeProjects = projectsWithStats.filter(
        (p: any) => p.statistics?.activitySummary?.lastMonth > 0,
      ).length;

      return {
        team: projects.team,
        summary: {
          totalProjects,
          totalFiles,
          activeProjects,
          activeProjectsPercentage:
            totalProjects > 0 ? (activeProjects / totalProjects) * 100 : 0,
        },
        projects: projects.projects,
        recentFiles: recentFiles?.files || [],
        metadata: {
          generatedAt: new Date().toISOString(),
          includeStats,
          includeRecentFiles,
          recentFilesLimit,
        },
      };
    } catch (error: any) {
      this.logger.error("Failed to generate team overview", {
        teamId,
        error: error.message,
      });
      throw error;
    }
  }

  /**
   * Quick project lookup by name
   * @param {string} teamId - Team ID
   * @param {string} projectName - Project name to find
   * @param {object} [options={}] - Search options
   * @returns {Promise<object|null>} Project data or null if not found
   */
  async findProject(
    teamId: string,
    projectName: string,
    options: Record<string, any> = {},
  ): Promise<Record<string, any> | null> {
    try {
      const searchResults = await this.searchProjects(teamId, projectName, {
        exactMatch: true,
        caseSensitive: options.caseSensitive || false,
      });

      if (searchResults.totalMatches === 0) {
        return null;
      }

      const project = searchResults.results[0];

      // Get project files if requested
      if (options.includeFiles) {
        const filesData = await this.getProjectFiles(project.id);
        project.files = filesData.files;
        project.fileCount = filesData.totalCount;
      }

      return project;
    } catch (error: any) {
      this.logger.error("Failed to find project", {
        teamId,
        projectName,
        error: error.message,
      });
      throw error;
    }
  }

  /**
   * Bulk file search across all projects
   * @param {string} teamId - Team ID
   * @param {string} fileName - File name to search for
   * @param {object} [options={}] - Search options
   * @returns {Promise<object[]>} Array of matching files with project context
   */
  async findFileAcrossProjects(
    teamId: string,
    fileName: string,
    options: Record<string, any> = {},
  ): Promise<Record<string, any>> {
    try {
      this.logger.debug("Searching for file across projects", {
        teamId,
        fileName,
        options,
      });

      const projectsWithFiles = await this.getProjectTree(teamId);
      const matchingFiles: any[] = [];

      const searchTerm = options.caseSensitive
        ? fileName
        : fileName.toLowerCase();

      for (const project of projectsWithFiles.projects) {
        for (const file of project.files) {
          const fileNameToSearch = options.caseSensitive
            ? file.name
            : file.name.toLowerCase();

          const matches = options.exactMatch
            ? fileNameToSearch === searchTerm
            : fileNameToSearch.includes(searchTerm);

          if (matches) {
            matchingFiles.push({
              ...file,
              projectId: project.id,
              projectName: project.name,
            });
          }
        }
      }

      return {
        query: {
          fileName,
          teamId,
          caseSensitive: options.caseSensitive || false,
          exactMatch: options.exactMatch || false,
        },
        results: matchingFiles,
        totalMatches: matchingFiles.length,
        projectsScanned: projectsWithFiles.totalProjects,
        totalFilesScanned: projectsWithFiles.totalFiles,
        searchedAt: new Date().toISOString(),
      };
    } catch (error: any) {
      this.logger.error("Failed to search file across projects", {
        teamId,
        fileName,
        error: error.message,
      });
      throw error;
    }
  }
}

export default FigmaProjectsSDK;
